packs/chef/actions/lib/shellhelpers.py

```diff
@@ -0,0 +1,62 @@
+import argparse
+import sys
+import os
+import subprocess
+import select
+import io
+
+# Starts command in subprocess processing output as it appears.
+def shell_out(command, env={}):
+    env = os.environ.copy().update(env)
+    shell = False if isinstance(command, list) else True
+    proc = subprocess.Popen(args=command, stdin=None, stdout=subprocess.PIPE,
+                               stderr=subprocess.PIPE, env=env, shell=shell)
+    while True:
+        fdlist = [proc.stdout.fileno(), proc.stderr.fileno()]
+        ios = select.select(fdlist, [], [])
+        for fd in ios[0]:
+            if fd == proc.stdout.fileno():
+                sys.stdout.write(proc.stdout.read(io.DEFAULT_BUFFER_SIZE))
+            else:
+                sys.stderr.write(proc.stderr.read(io.DEFAULT_BUFFER_SIZE))
+
+        if proc.poll() != None: break
+    return proc.returncode
+
+
+class CmdlineParser(object):
+    def __init__(self, parser_options):
+        '''
+        Initialize argument parser with argument triplets as the following:
+            [
+                ('-j', '--attributes', {}),
+                ...
+                ('-W', '--why_run', {'action': 'store_true'}),
+            ]
+        '''
+        self.parser = argparse.ArgumentParser()
+        self._keyname = {}
+        for k, lg, kwargs in parser_options:
+            self.parser.add_argument(k, lg , **kwargs)
+            self._keyname[ lg.lstrip('-') ] = k
+
+    def parse(self, argv=[]):
+        return vars( self.parser.parse_args(argv or None) )
+
+    def short_arglist(self, kwargs={}):
+        _arg_list(kwargs or self.parse(), short=True)
+
+    def long_arglist(self, kwargs={}):
+        _arg_list(kwargs or self.parse(), short=False)
+
+    def _arg_list(self, kwargs, short=None):
+        '''
+        Returns list of command line arguments with short or long key names.
+        '''
+        cmd = []
+        for n, v in (self.args or self.parse()).items():
```

Comment:
I haven't been thinking about caching it though. In case, say it, I want to run short_arglist several time with different kwargs. But it's had no use finally...